File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/rabbitmq/docs/consumer/acknowledge/type/ProductListener.kt
package io.micronaut.rabbitmq.docs.consumer.acknowledge.type

import io.micronaut.context.annotation.Requires
// tag::imports[]
import io.micronaut.rabbitmq.annotation.Queue
import io.micronaut.rabbitmq.annotation.RabbitListener
import io.micronaut.rabbitmq.bind.RabbitAcknowledgement
import java.util.concurrent.atomic.AtomicInteger
// end::imports[]

@Requires(property = "spec.name", value = "AcknowledgeSpec")
// tag::clazz[]
@RabbitListener
class ProductListener {

    val messageCount = AtomicInteger()

    @Queue(value = "product") // <1>
    fun receive(data: ByteArray, acknowledgement: RabbitAcknowledgement) { // <2>
        val count = messageCount.getAndUpdate { intValue -> intValue + 1 }
        if (count == 0) {
            acknowledgement.nack(false, true) // <3>
        } else if (count > 3) {
            acknowledgement.ack(true) // <4>
        }
    }
}
// end::clazz[]

File: docs-examples/example-python/src/test/python/micronaut/rabbitmq/docs/consumer/acknowledge/type/ProductListener.py
from micronaut.context.annotation import Requires
# tag::imports[]
from micronaut.rabbitmq.annotation import Queue, RabbitListener
from micronaut.rabbitmq.bind import RabbitAcknowledgement
# end::imports[]


@Requires(property="spec.name", value="AcknowledgeSpec")
# tag::clazz[]
@RabbitListener
class ProductListener:

    def __init__(self) -> None:
        self.message_count = 0

    @Queue(value="product")  # <1>
    def receive(self, data: bytes, acknowledgement: RabbitAcknowledgement) -> None:  # <2>
        count = self.message_count
        self.message_count += 1
        if count == 0:
            acknowledgement.nack(False, True)  # <3>
        elif count > 3:
            acknowledgement.ack(True)  # <4>
# end::clazz[]
